File: src/ReunionPhotos.Web/Modules/OneDriveSync/GraphApiClient.cs
using System.Net.Http.Json;
using System.Text.Json;
using Microsoft.Extensions.Options;
using ReunionPhotos.Web.Services;

namespace ReunionPhotos.Web.Modules.OneDriveSync;

// Talks to Microsoft Graph for the personal OneDrive account. Handles the
// token refresh dance, including the fact that personal-account refresh
// tokens rotate on every use - each refresh call returns a brand new
// refresh token, and the old one stops working. We persist the latest one
// in Airtable so this survives app restarts and redeploys.
public class GraphApiClient
{
    private const string GraphBase = "https://graph.microsoft.com/v1.0";
    private const string ConfigKeyRefreshToken = "OneDriveRefreshToken";

    private readonly HttpClient _http;
    private readonly GraphOptions _options;
    private readonly AirtableService _airtable;
    private readonly ILogger<GraphApiClient> _logger;

    private string? _accessToken;
    private DateTimeOffset _accessTokenExpiresAt = DateTimeOffset.MinValue;

    public GraphApiClient(HttpClient http, IOptions<GraphOptions> options, AirtableService airtable, ILogger<GraphApiClient> logger)
    {
        _http = http;
        _options = options.Value;
        _airtable = airtable;
        _logger = logger;
    }

    public async Task<List<DriveItemDto>> ListFolderItemsAsync(string driveId, string folderId, CancellationToken ct = default)
    {
        var token = await GetAccessTokenAsync(ct);
        var url = $"{GraphBase}/drives/{driveId}/items/{folderId}/children" +
                  "?$select=id,name,file,photo,image,createdDateTime";

        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", token);
        using var response = await _http.SendAsync(request, ct);
        response.EnsureSuccessStatusCode();

        var payload = await response.Content.ReadFromJsonAsync<DriveChildrenResponse>(cancellationToken: ct);
        return (payload?.Value ?? new())
            .Where(i => i.File is not null) // skip subfolders
            .ToList();
    }

    // Streams the actual image bytes for a photo. Used by the /image/{photoId}
    // proxy endpoint so the browser never needs a direct (short-lived) OneDrive
    // download URL - it always asks our server, which fetches fresh each time.
    public async Task<(Stream Content, string ContentType)> GetImageStreamAsync(
        string driveId, string itemId, bool thumbnail, CancellationToken ct = default)
    {
        var token = await GetAccessTokenAsync(ct);
        var url = thumbnail
            ? $"{GraphBase}/drives/{driveId}/items/{itemId}/thumbnails/0/large/content"
            : $"{GraphBase}/drives/{driveId}/items/{itemId}/content";

        var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", token);
        var response = await _http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, ct);
        response.EnsureSuccessStatusCode();

        var contentType = response.Content.Headers.ContentType?.MediaType ?? "image/jpeg";
        var stream = await response.Content.ReadAsStreamAsync(ct);
        return (stream, contentType);
    }

    private async Task<string> GetAccessTokenAsync(CancellationToken ct)
    {
        if (_accessToken is not null && DateTimeOffset.UtcNow < _accessTokenExpiresAt)
            return _accessToken;

        var tokenEndpoint = $"https://login.microsoftonline.com/{_options.TenantId}/oauth2/v2.0/token";

        // Prefer the most recently persisted refresh token (it may have
        // rotated since the app started); fall back to the configured seed
        // value the very first time the app runs.
        var refreshToken = await _airtable.GetConfigValueAsync(ConfigKeyRefreshToken, ct) ?? _options.RefreshToken;

        var form = new Dictionary<string, string>
        {
            ["client_id"] = _options.ClientId,
            ["client_secret"] = _options.ClientSecret,
            ["grant_type"] = "refresh_token",
            ["refresh_token"] = refreshToken,
            ["scope"] = "Files.Read offline_access",
        };

        using var response = await _http.PostAsync(tokenEndpoint, new FormUrlEncodedContent(form), ct);
        var body = await response.Content.ReadAsStringAsync(ct);

        if (!response.IsSuccessStatusCode)
        {
            _logger.LogError("OneDrive token refresh failed: {Body}", body);
            throw new InvalidOperationException("Failed to refresh Microsoft Graph access token. See logs for details.");
        }

        using var doc = JsonDocument.Parse(body);
        var root = doc.RootElement;
        _accessToken = root.GetProperty("access_token").GetString();
        var expiresIn = root.GetProperty("expires_in").GetInt32();
        _accessTokenExpiresAt = DateTimeOffset.UtcNow.AddSeconds(expiresIn - 60); // refresh a minute early

        // Persist the new refresh token - personal Microsoft accounts issue
        // a new one on every refresh and invalidate the old one.
        if (root.TryGetProperty("refresh_token", out var newRefreshTokenEl))
        {
            var newRefreshToken = newRefreshTokenEl.GetString();
            if (!string.IsNullOrEmpty(newRefreshToken))
                await _airtable.SetConfigValueAsync(ConfigKeyRefreshToken, newRefreshToken, ct);
        }

        return _accessToken!;
    }

    private class DriveChildrenResponse
    {
        public List<DriveItemDto> Value { get; set; } = new();
    }
}

public class DriveItemDto
{
    public string Id { get; set; } = "";
    public string Name { get; set; } = "";
    public object? File { get; set; }
    public DateTimeOffset? CreatedDateTime { get; set; }
}

File: src/modules/oneDriveSync/graphApiClient.js
// Talks to Microsoft Graph for a work/school OneDrive account. Handles the
// token refresh dance, including the fact that refresh tokens rotate on
// every use - each refresh call returns a brand new refresh token, and the
// old one stops working. We persist the latest one in Airtable so this
// survives app restarts and redeploys.
const config = require('../../config');
const airtable = require('../../services/airtableService');

const GRAPH_BASE = 'https://graph.microsoft.com/v1.0';
const CONFIG_KEY_REFRESH_TOKEN = 'OneDriveRefreshToken';

let accessToken = null;
let accessTokenExpiresAt = 0;

async function getAccessToken() {
  if (accessToken && Date.now() < accessTokenExpiresAt) {
    return accessToken;
  }

  const tokenEndpoint = `https://login.microsoftonline.com/${config.graph.tenantId}/oauth2/v2.0/token`;

  // Prefer the most recently persisted refresh token (it may have rotated
  // since the app started); fall back to the configured seed value the
  // very first time the app runs.
  const storedToken = await airtable.getConfigValue(CONFIG_KEY_REFRESH_TOKEN);
  const refreshToken = storedToken || config.graph.refreshToken;

  const form = new URLSearchParams({
    client_id: config.graph.clientId,
    client_secret: config.graph.clientSecret,
    grant_type: 'refresh_token',
    refresh_token: refreshToken,
    scope: 'Files.Read offline_access',
  });

  const res = await fetch(tokenEndpoint, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: form,
  });

  const body = await res.json();

  if (!res.ok) {
    console.error('OneDrive token refresh failed:', body);
    throw new Error('Failed to refresh Microsoft Graph access token. See logs for details.');
  }

  accessToken = body.access_token;
  accessTokenExpiresAt = Date.now() + (body.expires_in - 60) * 1000; // refresh a minute early

  // Persist the new refresh token - it rotates on every use and invalidates
  // the previous one.
  if (body.refresh_token) {
    await airtable.setConfigValue(CONFIG_KEY_REFRESH_TOKEN, body.refresh_token);
  }

  return accessToken;
}

async function listFolderItems(driveId, folderId) {
  const token = await getAccessToken();
  const url = `${GRAPH_BASE}/drives/${driveId}/items/${folderId}/children?$select=id,name,file,photo,image,createdDateTime`;

  const res = await fetch(url, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!res.ok) {
    throw new Error(`Graph list folder items failed: ${res.status} ${await res.text()}`);
  }

  const body = await res.json();
  return (body.value || []).filter((item) => !!item.file); // skip subfolders
}

// Streams the actual image bytes for a photo. Used by the /image/:photoId
// proxy route so the browser never needs a direct (short-lived) OneDrive
// download URL - it always asks our server, which fetches fresh each time.
async function getImageStream(driveId, itemId, thumbnail) {
  const token = await getAccessToken();
  const url = thumbnail
    ? `${GRAPH_BASE}/drives/${driveId}/items/${itemId}/thumbnails/0/large/content`
    : `${GRAPH_BASE}/drives/${driveId}/items/${itemId}/content`;

  const res = await fetch(url, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!res.ok) {
    throw new Error(`Graph get image failed: ${res.status} ${await res.text()}`);
  }

  return {
    body: res.body, // a web ReadableStream, pipeable straight into the Express response
    contentType: res.headers.get('content-type') || 'image/jpeg',
  };
}

module.exports = { listFolderItems, getImageStream };
